Warm workers docs (#6788)
PipedreamHQ/pipedream: docs/docs/workflows/settings/README.md

[Manage the concurrency and rate](/workflows/concurrency-and-throttling/) at which events from a source trigger your workflow code.

## Eliminate cold starts

> If your workflow doesn't process an event for roughly 5 minutes, Pipedream turns off the execution environment that runs your code. When your workflow receives another event, Pipedream creates a new execution environment to process your event. **Initializing this environment takes a few seconds, which delays the execution of this first event**. This is common on serverless platforms, and is typically referred to as a "cold start".

Author: ```suggestion
If your workflow doesn't process an event for about 5 minutes, Pipedream turns off the execution environment that runs your code. When your workflow receives another event, Pipedream creates a new execution environment to process your event. **Initializing this environment takes a few seconds, which delays the execution of this first event**. This is common on serverless platforms and is typically referred to as a "cold start".
```

Author: Changed "roughly" to "about" to follow Google's style guide for clarity.

> You can reduce cold starts by configuring a number of dedicated **workers**:

Author: ```suggestion
You can reduce cold starts by configuring a number of dedicated **workers**:
```

Author: Removed extra newline to follow Dreyer's English for consistency.

Pipedream also performs some initialization operations on new workflow runs, so you may still observe a small startup time (typically around 50ms per workflow step) on dedicated workers.

:::

### Pricing for dedicated workers

You're charged {{$site.themeConfig.WARM_WORKERS_CREDITS_PER_INTERVAL}} credits for every {{$site.themeConfig.WARM_WORKERS_INTERVAL}} a dedicated worker is live, per worker, per {{$site.themeConfig.MEMORY_LIMIT}} memory. You can view the credits used by dedicated workers [in your billing settings](https://pipedream.com/settings/billing):

<div>
<img src="https://res.cloudinary.com/pipedreamin/image/upload/v1686015101/docs/Screenshot_2023-06-05_at_6.26.53_PM_stlljl.png" alt="Warm workers credit usage in billing" />
</div>

For example, if you run a single dedicated worker for 24 hours, that would cost 720 credits:

```
5 credits per 10 min
* 6 10-min periods per hour
* 24 hours
= 720 credits
```

{{$site.themeConfig.WARM_WORKERS_INTERVAL}} is the _minimum_ interval that Pipedream charges for usage. If you have a dedicated worker live for 1 minute, Pipedream will still charge {{$site.themeConfig.WARM_WORKERS_CREDITS_PER_INTERVAL}} credits.

> 

Author: ```suggestion
You should configure dedicated workers when you need to process requests as soon as possible, with no latency. For example, you may build an HTTP-triggered workflow that returns a synchronous HTTP response to a user. These workflows need to respond quickly.
```

Additionally, any change to dedicated worker configuration, (including worklow deploys) will result in an extra {{$site.themeConfig.WARM_WORKERS_CREDITS_PER_INTERVAL}} credits of usage.

> 

Author: ```suggestion
### How many workers should I configure?
```

Author: Removed extra newline to follow Dreyer's English for consistency.

<div>
<img src="https://res.cloudinary.com/pipedreamin/image/upload/v1686079387/docs/Screenshot_2023-06-06_at_12.23.03_PM_mpsqek.png" alt="Dedicated worker overlap" />
</div>

> 

Author: ```suggestion
But you might have a higher-volume app that receives two concurrent requests. In that case, Pipedream spins up **two** workers to handle each request.
```

Author: Changed ""But"" to ""However"" to follow Google's style guide for clarity."

### When should I configure dedicated workers?

> 

Author: ```suggestion
For many user-facing (even internal) applications, the number of requests over time can be modeled with a [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution). You can use that distribution to estimate the number of workers you need at an average time, or set it higher if you want to ensure most requests hit a dedicated worker. You can also save a record of all workflow runs to your own database, with the timestamp they ran ([see `steps.trigger.context.ts`](/workflows/events/#steps-trigger-context)), and look at your own pattern of requests, to compute the optimal number of workers.
```

Author: Changed ""distrubution"" to ""distribution"" to correct spelling error."

You should configure dedicated workers when you need to process requests as soon as possible, with no latency. For example, you may build an HTTP-triggered workflow that returns a synchronous HTTP response to a user. These workflows need to respond quickly.

### How many workers should I configure?

Incoming requests are handled by a single worker, one at a time. If you only receive one request a minute, and the workflow finishes execution in a few seconds, you may only need one worker.

But you might have a higher-volume app that receives two concurrent requests. In that case, Pipedream spins up **two** workers to handle each request.

> For many user-facing (even internal) applications, the number of requests over time can be modeled with a [Poisson distrubution](https://en.wikipedia.org/wiki/Poisson_distribution). You can use that distribution to estimate the number of workers you need at an average time, or set it higher if you want to ensurea specific percentage of requests hit a dedicated worker. You can also save a record of all workflow runs to your own database, with the timestamp they ran ([see `steps.trigger.context.ts`](/workflows/events/#steps-trigger-context)), and look at your own pattern of requests, to compute the optimal number of workers.

Author: ```suggestion
For many user-facing (even internal) applications, the number of requests over time can be modeled with a [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution). You can use that distribution to estimate the number of workers you need at an average time, or set it higher if you want to ensure a specific percentage of requests hit a dedicated worker. You can also save a record of all workflow runs to your own database, with the timestamp they ran ([see `steps.trigger.context.ts`](/workflows/events/#steps-trigger-context)), and look at your own pattern of requests, to compute the optimal number of workers.
```

Author: Changed ""ensurea"" to ""ensure a"" to correct spelling error.
